rfd/0162-machine-id-token-join-method-bot-instance.md

```diff
@@ -0,0 +1,565 @@
+one-to-one binding between a single instance of `tbot` and a single Bot user.
+This is not the case when using the delegated join methods.
+ 
+### Persistent Bot Instance Identity
+
+Today, there is no persistent identifier for an individual instance of a Bot.
+Instead, all `tbot` instances are effectively identified solely by their Bot
+identity. There is no easy way to distinguish them. On each renewal, `tbot`
+regenerates the private-public key pair that is used within its certificate and
+there is no other form of unique ID that is persisted across renewals.
+
+This poses a few challenges:
+
+- For the purposes of auditing, it is not possible to trace actions to a
+  specific instance of a Bot.
+- For improving the `token` join method to support multiple Bot instances
+  associated with a single Bot, there is no identifier to correlate with the 
+  generation counter.
+- For analytics purposes, it's difficult for us to track the number of
+  individual Bot instances in use. We cannot easily determine if it's a single
+  very active Bot instance, or many less active Bot instances.
+
+To rectify this, a unique identifier should be established for an instance of
+a Bot.
+
+#### Public Key Fingerprint
+
+One option is to modify the behaviour of `tbot` to persist and reuse the
+keypair across renewals. We could then use a fingerprint of the public key as a
+unique identifier of the Bot instance.
+
+This feels like a natural identifier. It avoids introducing a new attribute to
+certificates as the public key is already encoded within certificates. The
+nature of public-key cryptography also means that this provides the Bot instance
+a way to identify itself without needing an issued certificate.
+
+To mitigate the risk of pre-image attacks, SHA256 will be used to determine the
+fingerprint of the public key. In addition, the full public key should be
+recorded and verified against when authenticating a Bot instance action.
+
+It should be noted that with this technique, rotating the keypair of a `tbot`
+instance would reset the identity of that instance. Rotation of this keypair
+would be unusual and this side effect seems expected.
+
+##### Alternative: UUID Certificate Attribute
+
+On the initial join of a Bot instance, we could generate a UUID to identify that 
+Bot instance and encode this within the certificate. Upon renewals, the UUID
+would be copied from the current certificate and into the new one.
+
+Whilst this is fairly easy to implement for the `token` join method, one
+challenge for the delegated join methods is that rather than renewing, the
+`tbot` instance merely re-joins. As the join RPCs are unauthenticated, the 
+previous certificate of the Bot instance is not readily available. We can
+either:
+
+- Accept this limitation and treat each renewal of a delegated Bot instance as a
+  new Bot instance. This is likely unacceptable and would limit any advantages
+  of this work to the `token` join method.
```

Comment:
> I'd expect all delegated joins from CI providers to be new bot instances every time, since they're stateless.

Hmm, I'm not sure I'd necessarily want to set that precedent. Even in CI situations, a long-running CI job may renew several times (e.g a background tbot database tunnel with a CI job running a long migration or an ansible job connecting to hundreds of hosts to roll out an update) and I think the events associated with that long-running job being associated with different bot instances would be confusing. 

> And I imagine a decent share of the cloud provider / k8s joins that could be stateful end up being ephemeral anyway.

Right now, there's not much of a motivation to have state configured because there's no real advantage to it, in fact, at one point there was a disadvantage as an expired cert would cause the bot to be unable to start. I'd imagine that if we were to introduce BotInstance's then there's a perceived benefit at least in being able to attribute actions to a particular instance of tbot throughout its lifetime in a fairly persistent way.

I suppose the other overriding thought in my head is I'm desperate to stop the token vs delegated join method behaviour diverting further. It's already confusing enough and I can see user frustration when the less secure join method is actually delivering more features to them (in terms of having a long-lived persistence identity they can attribute actions to).